src/NATS.Client.Core/Commands/ProtocolWriter.cs

```diff
@@ -48,70 +48,131 @@ public void WritePong()
+        }
+        else
+        {
+            // 'HPUB '  + subject                                +' '+ header len         +' '+ payload len        +'\r\n'
+            ctrlLen = 5 + _subjectEncoding.GetByteCount(subject) + 1 + MaxIntStringLength + 1 + MaxIntStringLength + 2;
         }
-
-        // Start writing the message to buffer:
-        // PUP / HPUB
-        _writer.WriteSpan(headers == null ? CommandConstants.PubWithPadding : CommandConstants.HPubWithPadding);
-        _writer.WriteASCIIAndSpace(subject);
 
         if (replyTo != null)
         {
-            _writer.WriteASCIIAndSpace(replyTo);
+            // len  += replyTo                                +' '
+            ctrlLen += _subjectEncoding.GetByteCount(replyTo) + 1;
         }
 
+        var ctrlSpan = _writer.GetSpan(ctrlLen);
+        var span = ctrlSpan;
         if (headers == null)
         {
-            _writer.WriteNumber(payload.Length);
+            span[0] = (byte)'P';
+            span[1] = (byte)'U';
+            span[2] = (byte)'B';
+            span[3] = (byte)' ';
+            span = span[4..];
         }
         else
         {
-            var headersLength = _bufferHeaders.WrittenSpan.Length;
-            _writer.WriteNumber(CommandConstants.NatsHeaders10NewLine.Length + headersLength);
-            _writer.WriteSpace();
-            var total = CommandConstants.NatsHeaders10NewLine.Length + headersLength + payload.Length;
-            _writer.WriteNumber(total);
+            span[0] = (byte)'H';
+            span[1] = (byte)'P';
+            span[2] = (byte)'U';
+            span[3] = (byte)'B';
+            span[4] = (byte)' ';
+            span = span[5..];
         }
 
-        // End of message first line
-        _writer.WriteNewLine();
+        var written = _subjectEncoding.GetBytes(subject, span);
+        span[written] = (byte)' ';
+        span = span[(written + 1)..];
 
-        if (headers != null)
+        if (replyTo != null)
         {
-            _writer.WriteSpan(CommandConstants.NatsHeaders10NewLine);
-            _writer.WriteSpan(_bufferHeaders.WrittenSpan);
+            written = _subjectEncoding.GetBytes(replyTo, span);
+            span[written] = (byte)' ';
+            span = span[(written + 1)..];
         }
 
-        if (payload.Length != 0)
+        Span<byte> lenSpan;
+        if (headers == null)
         {
-            _writer.WriteSequence(payload);
+            // len =            payload len
+            lenSpan = span[..MaxIntStringLength];
+            span = span[lenSpan.Length..];
+        }
+        else
+        {
+            // len =          header len         +' '+ payload len
+            lenSpan = span[..(MaxIntStringLength + 1 + MaxIntStringLength)];
+            span = span[lenSpan.Length..];
         }
 
-        _writer.WriteNewLine();
-    }
+        span[0] = (byte)'\r';
+        span[1] = (byte)'\n';
```

Comment:
If we do `Int` for others, maybe `short` here?

Reply:
Setting the index directly is actually faster for 2 bytes than `BinaryPrimitives.WriteUInt16LittleEndian`, see [gist](https://gist.github.com/caleblloyd/bcfcbdee77757c54c00a226082debb4b)